src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Dirac mixture single publisher model."""
+
+import numpy as np
+import cvxpy as cp
+from scipy import stats
+from typing import List, Callable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+
+class UnivariateMixedPoissonOptimizer:
+    """Fit an univariate mixed Poisson distribution on a frequency histogram.
+
+    A mixed Poisson distribution has probablity mass function (pmf)
+    f(x) = sum_{j=1}^J w_j pi(x | lambda_j), x = 0, 1, 2, ...
+    where any
+    pi(x | lambda) = lambda^x * exp(-lambda) / x!
+    is the pmf of Poisson distribution.  Each lambda_j is called a component.
+    w_j is called the weight of component lambda_j.
+
+    At this moment, the fitting algorithm consists of a grid search of components
+    and a convex optimization on the weights of different components.
+    """
+
+    def __init__(
+        self,
+        frequency_histogram: np.ndarray,
+        ncomponents: int = 200,
```

Comment:
I wonder if it would be helpful to use an L1-penalty to identify the true number of non-zero components?

Reply:
@matthewclegg  Can you please elaborate what you mean?

Reply:
@amiragha IIUC, Matt thinks it'll be helpful to have sparsity of weights to identify the true number of non-zero components.  However, the cross-entropy criterion in this PR does not lead to sparse weights.  In order to have sparsity, one way is to add an L1 penalty to the cross-entropy criterion.